UnitTests/GravatarTests/MruImageCacheTests.cs

```diff
@@ -0,0 +1,254 @@
+﻿using System;
+using System.Drawing;
+using System.Threading;
+using System.Threading.Tasks;
+using Gravatar;
+using NSubstitute;
+using NUnit.Framework;
+
+namespace GravatarTests
+{
+    [TestFixture]
+    public class MruImageCacheTests : IDisposable
+    {
+        private readonly Image _img1 = new Bitmap(16, 16);
+        private readonly Image _img2 = new Bitmap(16, 16);
+        private readonly Image _img3 = new Bitmap(16, 16);
+        private readonly Image _img4 = new Bitmap(16, 16);
+
+        private IImageCache _inner;
+        private IImageCache _cache;
+
+        [SetUp]
+        public void SetUp()
+        {
+            _inner = Substitute.For<IImageCache>();
```

Comment:
❗️ Do not use `Substitute` in new tests. Use one of the following alternatives instead (in order of preference):

1. A production type, e.g. `DirectoryImageCache`
2. A shared concrete type defined in test code, e.g. `FakeInMemoryImageCache` that is implemented according to the known API of the underlying interface
3. A specialized type defined in this test class - this is often a derivation of a type described in the previous item but with one or more methods overridden to provide special behavior for testing an edge case

Reply:
Why not? For these kinds of tests it's very useful.

In my pending PR around git executable management I have written a custom mock because the operations are more easily expressed with it in that case. Here the generated one does fine for my taste.

Reply:
> Why not?

There is a long discussion about this in another topic, but it boils down to this:

1. The substitute does not automatically adhere to the pre- and post-conditions of methods in the interface
2. The compiler provides no assistance to point out cases where a substitute type violates a pre- or post-condition (e.g. there is no compile error, and there is no ability to Go To Implementation)
3. Substitute types do not represent the true behavior of an application, allowing tests to be misleading
4. In practice, type substitution increases the maintenance costs of an application by unnecessarily coupling observable behavior of objects with their implementation details
5. We are not trying to support a stable+extensible+public API, which is the one case where substitute types have value

> For these kinds of tests it's very useful.

The problems here do not surround the use of a substitute type for the creation of a test. The problems appear in long-term maintenance of the application and test suite.